### PR TITLE
Implement ChainStorageService

### LINKED ISSUE #80: Create initial ChainStoreService
Create ChainStoreService.java:
- stores blocks, attestations, and states
- controls what to keep in memory and what to write to disk

> Note: Alway worth looking at what Pantheon did:
> https://github.com/PegaSysEng/pantheon/tree/master/services/kvstore/src/main/java/tech/pegasys/pantheon/services/kvstore

### PR DESCRIPTION
**Pull request by schroedingerscode**
## PR Description
Implement a chain storage service to intelligently persist/cache new blocks and attestations from the p2p network, and relay to relevant parties such as StateTreeManager on request.

**Tasks**
- [x] **Pull in Cava kvstore and create chain storage service implementation class**
- [x] **Subscribe chain storage implementation to (mock) p2p events**
- [x] **Store Attestations**
- [x] **Store BeaconBlocks**
_(This requires creating constructors and might as well changes primitive arrays to Lists while we're at it.)_

I've asked to merge these changes in as we work on catching up with the Spec v0.1, to cut down on merge conflicts later. An additional PR will follow in time.

### CODE REVIEW

File: ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/blocks/BeaconBlock.java [hunk 1]
```diff
@@ -22,26 +23,41 @@
...
 
   // Body
   private BeaconBlockBody body;
 
-  public BeaconBlock() {}
+  public BeaconBlock(
+      long slot,
+      List<Bytes32> ancestor_hashes,
+      Bytes32 state_root,
+      Bytes32 randao_reveal,
```
Comment (jrhea) [inline ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/blocks/BeaconBlock.java]: Bytes32 randao_reveal should be List<Bytes48> randao_reveal

**Review by jrhea — COMMENTED** [review]
Another thing to consider is:

StateTransitionManager. java needs to remove:
```
@Subscribe
  public void onNewBlock(BeaconBlock block) {
    LOG.info("New Beacon Block Event detected");
    block.setSlot(this.state.getSlot());
    unprocessedBlocks.add(block);
  }

  @Subscribe
  public void onNewAttestation(Attestation attestation) {
    LOG.info("New Attestation Event detected");
    unprocessedAttestations.add(attestation);
  }
```
I see those events are handles in ChainStaorge.java now, which is good; however, StateTreeManager.java will now need a way to request the newBlocks and Attestations from ChainStore.  This sounds like an unnecessary a abstraction, but it will pay off later when the chainstore handles when to cache to disk, prune, etc.

**Comment by schroedingerscode** [comment]
@jrhea - Concerning the StateTreeManager also subscribing to the same event, yep I totally agree with you. 

I've got this WIP PR open for people to track and comment, however I'm working a phased approach to implementation that I've poorly attempted to express and track with the checkboxes in the description. 

Once I create the relevant callbacks for StateTreeManager to request updates from ChainStorageService, I'll sever the subscription connection between StateTreeManager and the (Mock) P2P Service. 

**tl;dr The concern above should be obsoleted by further commits to this WIP PR.**

**Review by jrhea — APPROVED** [review]
lgtm
